[Tan De Yi] iP

deyixtan opened the pull request.

Commit:
Add Gradle support

Commit:
build.gradle: Update version to 8.29

deyixtan committed:
Implement greeting, echo and exit

deyixtan committed:
Implement add and list

deyixtan committed:
Implement mark as done

deyixtan committed:
Implement to-dos, events and deadlines

deyixtan committed:
Set up automated text UI testing

deyixtan committed:
Fix bugs in automated text UI testing

- Remove unicode, and use simple characters to mark task as done
- Check for empty redirection input which could've resulted in NoSuchElementException

deyixtan committed:
Implement error handling

deyixtan committed:
Minor output changes

deyixtan committed:
Update error handling

- Checks if task already marked as done
- Able to identify commands better
- Abstracted some code (i.e. shorten code)

deyixtan committed:
Implement delete

deyixtan committed:
Update command parsing using Enum

deyixtan committed:
Update error handling

- Check for misuse of arguments
- Check whether task description and argument are empty
- Code housekeeping

deyixtan committed:
Implement usage

- Include usage in automated test

deyixtan committed:
Code housekeeping

deyixtan committed:
Implement exit confirmation

- More code housekeeping

deyixtan committed:
Update automated text UI testing

- Test on wider range of inputs, which include more edge cases

deyixtan committed:
Implement save

deyixtan committed:
Implement date parser

deyixtan committed:
Resolve merge conflict with branch 'branch-Level-8'

deyixtan committed:
Implement load

deyixtan committed:
Refactor code to extract out closely related code as classes

deyixtan committed:
Organise project into packages

deyixtan committed:
Add some task-related JUnit tests

deyixtan committed:
Set up files for artifact creation

deyixtan committed:
Implement find

deyixtan committed:
Comply with coding standard

deyixtan committed:
Merge branch 'branch-A-CodingStandard'

deyixtan committed:
Add JavaDoc comments

deyixtan committed:
Merge branch 'branch-A-JavaDoc'

godjuansan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    public void run() {
+        ui.showGreetings();
+        boolean isExit = false;
```
For better quality, should make the boolean isExit to be private attributes of Duke class

godjuansan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,54 @@
+    public static Command parse(String input) throws DukeException {
+        String[] inputArr = input.split(" ");
+        String taskType = inputArr[0].toLowerCase();
+        String taskDesc = "";
+        String taskArg = "";
```
Should not declare it to be empty string here

godjuansan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,54 @@
+        }
+
+        // Get argument
+        for (i = i + 1; i < inputArr.length; i++) {
+            if (!taskArg.equals(""))
+                taskArg += " ";
+            taskArg += inputArr[i];
+        }
```
This is also not a good practice

godjuansan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+        try {
+            lines = Files.readAllLines(this.filePath);
+            for (String line : lines) {
+                String[] parts = line.replace('|', ' ').split("[ ]+");
```
Should explain what happens here.

godjuansan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                    task = new Todo(parts[2]);
+                else if (parts[0].equals("D")) {
+                    task = new Deadline(parts[2], LocalDate.parse(parts[3]));
+                } else
+                    task = new Event(parts[2], LocalDate.parse(parts[3]));
+
```
Missing curly bracket

godjuansan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                String[] parts = line.replace('|', ' ').split("[ ]+");
+
+                Task task;
+                if (parts[0].equals("T"))
```
What is parts[0]? Abstract more
e.g. String command = parts[0];
if (command.equals("T")) {
    ...
}

godjuansan replied:
Also missing curly bracket

godjuansan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                } else
+                    task = new Event(parts[2], LocalDate.parse(parts[3]));
+
+                if (parts[1].equals("1"))
```
Curly bracket

godjuansan replied:
Abstract more. What is parts[1]?

godjuansan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                    sb.append('E');
+
+                // Status
+                sb.append(" | ");
+                if (task.getStatusIcon().equals("*"))
+                    sb.append("1");
+                else
+                    sb.append("0");
+
+                // Description
+                sb.append(" | ");
+                sb.append(task.getDescription());
```
After reading this part, only then will I understand that | is the delimiter that you put to differentiate each component of the command. Should explain it clearly in the comments

godjuansan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+     * @return Formatted output.
+     */
+    public String showList() {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("You got a total of " + this.collection.size() + " task(s).\n");
+        for (int i = 0; i < collection.size(); i++) {
+           sb.append(String.format("\t%d. %s\n", i + 1, collection.get(i)));
+        }
+
```
Should use UI class to print

godjuansan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+     * @return Formatted output.
+     */
+    public String findTasks(String keyword) {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("Here are the matching tasks in your list:\n");
+        for (int i = 0; i < collection.size(); i++) {
+            Task task = this.collection.get(i);
+            if (task.getDescription().contains(keyword)) {
+                sb.append(String.format("\t%d. %s\n", i + 1, collection.get(i)));
+            }
+        }
+        return sb.toString();
```
Same. Use UI class to print instead of doing it directly here.

godjuansan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+        } catch (DateTimeParseException e) {
+            throw new DukeException("Please specify a proper date... (Format: YYYY-MM-DD)");
+        }
+        return "Got it, I have added the task '" + args[0] + "' to your collection.";
```
print using the UI class

godjuansan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,165 @@
+     * @throws DukeException If invalid index specified.
+     */
+    public String deleteTask(String index) throws DukeException {
+        try {
+            int itemIdx = Integer.parseInt(index.split(" ")[0]) - 1;
+            Task task = this.collection.remove(itemIdx);
+            return "Task '" + task.getDescription() + "' has been deleted.";
```
Again, print using the UI class

godjuansan commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
Maybe we can separate LocalDate and LocalTime to make it better

godjuansan commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+     * @return Outcome of the action.
+     */
+    public Boolean markAsDone() {
+        if (this.isDone)
+            return false;
```
Curly bracket

godjuansan commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,159 @@
+    protected String[] commandArgs;
+    protected boolean toExit;
+
+    public enum CommandsEnum {
+        USAGE, LIST, FIND,
+        TODO, DEADLINE, EVENT,
+        DONE, DELETE, SAVE, BYE
+    }
```
Create other class named CommandsEnum and put this there

godjuansan commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,159 @@
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        try {
+            CommandsEnum ce = CommandsEnum.valueOf(this.commandType.toUpperCase());
```
variable name could be better

godjuansan reviewed (commented):
Overall design-wise is ok, some improvement could be made.

deyixtan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

deyixtan committed:
Resolve mainClassName in gradle build file

deyixtan committed:
Add configuration files for checkstyle

deyixtan committed:
Update source code to comply with checkstyle

deyixtan committed:
Implement GUI

deyixtan committed:
Add assertions before every command execution

deyixtan committed:
Enable assertions

deyixtan committed:
Refactor code to improve code quality

deyixtan committed:
Merge pull request #1 from deyixtan/branch-A-Assertions

Add assertions

deyixtan committed:
Merge branch 'master' into branch-A-CodeQuality

deyixtan committed:
Merge pull request #2 from deyixtan/branch-A-CodeQuality

Refactor code

deyixtan committed:
Add task list breakdown

deyixtan committed:
Abstract commands

deyixtan committed:
Update chat bot to use local date and time

deyixtan committed:
Improve Ui code

deyixtan committed:
Remove unnecessary comments

deyixtan committed:
Improve handling of error message

deyixtan committed:
Update command parsing with regex

deyixtan committed:
Update label layout property to be able to display more content

deyixtan committed:
Update code for housekeeping

deyixtan committed:
Abstract input parsing and file data loading methods

deyixtan committed:
Refactor task list

deyixtan committed:
Update commands and command response

deyixtan committed:
Refactor UI

deyixtan committed:
Organise files into a more extensible hierarchy

deyixtan committed:
Fix parsing issue when load file data

deyixtan committed:
Add missing Override annotation for usage and save command

deyixtan committed:
Refactor tasks

deyixtan committed:
Add undone command

deyixtan committed:
Remove javadocs temporarily

deyixtan committed:
Update some Ui variable name

deyixtan committed:
Remove unused methods in task list

deyixtan committed:
Remove Ui requirement for commands to be executed

deyixtan committed:
Refactor chat bot
- Remove more comments
- Abstract the different dialog boxes
- Abstract statistics functionality into a command
- Add more details for CommandResponse to return

deyixtan committed:
Update source code to end with newline and explicit referencing

deyixtan committed:
Organise files into a more extensible hierarchy

deyixtan committed:
Update chat bot to benefit from TaskList abstraction

deyixtan committed:
Fix issue of displaying wrong display picture

deyixtan committed:
Update chat bot name to Quackers

deyixtan committed:
Update GUI to fit duck theme

deyixtan committed:
Remove all tests temporarily

deyixtan committed:
Fix Ui grammar

deyixtan committed:
Polish GUI

deyixtan committed:
Update GUI main stage properties

deyixtan committed:
Add representative screenshot

deyixtan committed:
Set theme jekyll-theme-leap-day

deyixtan committed:
Add auto-save

deyixtan committed:
Remove unused save messages

deyixtan committed:
Add brief user guide

deyixtan committed:
Merge branch 'master' of https://github.com/deyixtan/ip

deyixtan committed:
Update main repository's readme

deyixtan committed:
Update code to address invalid command from empty input and deleting invalid index issue

deyixtan committed:
Update code of mark done and undone to addresss edge cases

deyixtan committed:
Add Javadocs for all files

deyixtan committed:
Update representative screenshot

deyixtan committed:
Update brief user guide

deyixtan committed:
Update files to conform to checkstyle guidelines

deyixtan committed:
Add some basic task-related JUnit tests

deyixtan committed:
Fix representative screenshot in main README.md

deyixtan committed:
Set theme jekyll-theme-tactile

deyixtan committed:
Fix spelling mistake in README

deyixtan committed:
Merge branch 'master' of https://github.com/deyixtan/ip

deyixtan committed:
Update README release link and show download links in web page

deyixtan committed:
Set theme jekyll-theme-midnight

deyixtan committed:
Merge branch 'master' of https://github.com/deyixtan/ip

# Conflicts:
#	docs/_config.yml

deyixtan committed:
Update github pages theme to just-the-docs

deyixtan committed:
Fix just-the-docs remote link

deyixtan committed:
Update github pages configuration file

deyixtan committed:
Update github pages markdown and configuration file

deyixtan committed:
Update github pages icon

deyixtan committed:
Fix github pages page icon

deyixtan committed:
Fix some minor Ui issues
